From patchwork Fri Mar 13 16:28:36 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Lorenzo Bianconi <lorenzo@kernel.org>
X-Patchwork-Id: 14474081
X-Patchwork-Delegate: kuba@kernel.org
Received: from smtp.kernel.org (aws-us-west-2-korg-mail-1.web.codeaurora.org
 [10.30.226.201])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.subspace.kernel.org (Postfix) with ESMTPS id 32A533ACF1A
	for <netdev@vger.kernel.org>; Fri, 13 Mar 2026 16:28:48 +0000 (UTC)
Authentication-Results: smtp.subspace.kernel.org;
 arc=none smtp.client-ip=10.30.226.201
ARC-Seal: i=1; a=rsa-sha256; d=subspace.kernel.org; s=arc-20240116;
	t=1773419328; cv=none;
 b=T3pyj5sNv4uF1dPSlz9O3cC7vLmcO374PeCnYLfgDg6WYKv2JsvGvPkrqtABuvu5wdoUeKvzqLqkZnNZiRQdTkNczveGn7zDvJiJHF2tShUs2Y6AhvWLwj7uVAfXqE/46NWmkhM1eRYWU3cpyVw2Z985KPnPBMUSlh7R/XVmFmg=
ARC-Message-Signature: i=1; a=rsa-sha256; d=subspace.kernel.org;
	s=arc-20240116; t=1773419328; c=relaxed/simple;
	bh=xz7HYdb1hOl3DJa7Hqbhm8rgb1ssw/IvqbgQXX1MSR0=;
	h=From:Date:Subject:MIME-Version:Content-Type:Message-Id:To:Cc;
 b=QFe1bs564s0h5j2fBn54WvrC4tyFjn70zPQBlmfWrnZGxP3GCXPfWP6P2hwJwJZCgWTHeIq3UuLdmVVnX+6XJ9qyQ4hRbk+d3Pmoejmepe5D7yGfryI7aKj5FVI4tpmgj1zsHl8AXyuyDAwwCIytYuYhCQXmZOwMzW9UGtKPLW8=
ARC-Authentication-Results: i=1; smtp.subspace.kernel.org;
 dkim=pass (2048-bit key) header.d=kernel.org header.i=@kernel.org
 header.b=mPTb0cvQ; arc=none smtp.client-ip=10.30.226.201
Authentication-Results: smtp.subspace.kernel.org;
	dkim=pass (2048-bit key) header.d=kernel.org header.i=@kernel.org
 header.b="mPTb0cvQ"
Received: by smtp.kernel.org (Postfix) with ESMTPSA id 8A123C2BC86;
	Fri, 13 Mar 2026 16:28:47 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=kernel.org;
	s=k20201202; t=1773419327;
	bh=xz7HYdb1hOl3DJa7Hqbhm8rgb1ssw/IvqbgQXX1MSR0=;
	h=From:Date:Subject:To:Cc:From;
	b=mPTb0cvQTn1Ic7HMQNrbdhW0F0J3mGvt9sPq0HU4KoFUVESW0gec6WNmKNGwbqNrG
	 Pc65+HAUvKC3l5t+ofY1hK1rrWJ8QCdl2rUICk/QaoEH1TEzvYubUcfDYB1UjkBdcJ
	 ug8twP4qhw87Y/gYxX5vQFOa74WP53kkp4FTazRDZSx2pchqGwyE7EFkw3UY5kT1lG
	 zXBqfdl7gpOEL5WkcchQbZgoIUaqs+f4p5NXZKoYB3NO/TKwk1QNWc+DHbWLsrFvwQ
	 3eqsOhSvzu4D/zwtlX0ZhWKuXT/XdguYwlG6bM/HFZ0L3RoiV0dBjtxCpSYOgKdgDY
	 fSwMeLXLdomAA==
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Fri, 13 Mar 2026 17:28:36 +0100
Subject: [PATCH net-next v2] net: airoha: select QDMA block according
 LAN/WAN configuration
Precedence: bulk
X-Mailing-List: netdev@vger.kernel.org
List-Id: <netdev.vger.kernel.org>
List-Subscribe: <mailto:netdev+subscribe@vger.kernel.org>
List-Unsubscribe: <mailto:netdev+unsubscribe@vger.kernel.org>
MIME-Version: 1.0
Message-Id: <20260313-airoha-qdma-lan-wan-mode-v2-1-7d577db6e40c@kernel.org>
X-B4-Tracking: v=1; b=H4sIAAAAAAAC/4WNQQ6CMBREr2L+2m9oS8C48h6GRUMH+BFabQliC
 He34QIuZvEymTcbJURBottpo4hFkgSfQZ9P1A7W92BxmUkXuiqMKthKDIPlt5ssj9bzJ2cKDlw
 ZZ7QGOnSG8vwV0cl6qB/kMbPHOlOTm0HSHOL3+FzU0f/XL4oVVyXaunSw11rfn4ge4yXEnpp93
 3/oxCE4ywAAAA==
X-Change-ID: 20260310-airoha-qdma-lan-wan-mode-63d322eefef3
To: Andrew Lunn <andrew+netdev@lunn.ch>,
 "David S. Miller" <davem@davemloft.net>, Eric Dumazet <edumazet@google.com>,
 Jakub Kicinski <kuba@kernel.org>, Paolo Abeni <pabeni@redhat.com>,
 Lorenzo Bianconi <lorenzo@kernel.org>
Cc: Simon Horman <horms@kernel.org>, linux-arm-kernel@lists.infradead.org,
 linux-mediatek@lists.infradead.org, netdev@vger.kernel.org
X-Mailer: b4 0.14.2
X-Patchwork-Delegate: kuba@kernel.org

Before this patch even GDM ports were assigned to QDMA0 while odd GDM
ports were using QDMA1, so, based on the DTS configuration, both QDMA0
and QDMA1 can theoretically receive traffic destinated to the host cpu
from LAN or WAN GDM ports.
Airoha folks reported the hw design assumes the LAN traffic destinated
to the host cpu is be forwarded to QDMA0 while traffic received on WAN
GDM port is managed by QDMA1. For this reason, select QDMA block according
to the GDM port LAN or WAN configuration:
- QDMA0 is used for GDM LAN devices
- QDMA1 is used for GDM WAN device

Assuming a device with three GDM ports, a typical configuration could be:
- MT7530 DSA switch -> GDM1 (eth0) -> QDMA0 (LAN traffic)
- External PHY      -> GDM2 (eth1) -> QDMA1 (WAN traffic)
- External PHY      -> GDM3 (eth2) -> QDMA0 (LAN traffic)

We can then bridge eth0 DSA port (lanX) with eth2 since they all tx/rx
LAN traffic.

Please note this patch introduces a change not visible to the user since
airoha_eth driver currently supports just the internal phy available via
the MT7530 DSA switch and there are no WAN interfaces officially supported
since PCS/external phy is not merged mainline yet (it will be posted with
following patches).

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
Reviewed-by: Simon Horman <horms@kernel.org>
---
Changes in v2:
- Update commit log.
- Link to v1: https://lore.kernel.org/r/20260310-airoha-qdma-lan-wan-mode-v1-1-64ec74dea872@kernel.org
---
 drivers/net/ethernet/airoha/airoha_eth.c | 18 ++++++++----------
 drivers/net/ethernet/airoha/airoha_eth.h |  1 +
 2 files changed, 9 insertions(+), 10 deletions(-)


---
base-commit: 8f921f61005450589c0bc1a941a5ddde21d9aed9
change-id: 20260310-airoha-qdma-lan-wan-mode-63d322eefef3

Best regards,

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1791,11 +1791,13 @@ static int airhoha_set_gdm2_loopback(str
 static int airoha_dev_init(struct net_device *dev)
 {
 	struct airoha_gdm_port *port = netdev_priv(dev);
-	struct airoha_qdma *qdma = port->qdma;
-	struct airoha_eth *eth = qdma->eth;
+	struct airoha_eth *eth = port->eth;
 	u32 fe_cpu_port;
 	u8 ppe_id;
 
+	/* QDMA0 is used for lan ports while QDMA1 is used for WAN ports */
+	port->qdma = &eth->qdma[!airoha_is_lan_gdm_port(port)];
+	port->dev->irq = port->qdma->irq_banks[0].irq;
 	airoha_set_macaddr(port, dev->dev_addr);
 
 	switch (port->id) {
@@ -1819,7 +1821,7 @@ static int airoha_dev_init(struct net_de
 		}
 		fallthrough;
 	default: {
-		u8 qdma_id = qdma - &eth->qdma[0];
+		u8 qdma_id = port->qdma - &eth->qdma[0];
 
 		/* For PPE1 select cpu port according to the running QDMA. */
 		fe_cpu_port = qdma_id ? FE_PSE_PORT_CDM2 : FE_PSE_PORT_CDM1;
@@ -3014,11 +3016,10 @@ static int airoha_setup_phylink(struct n
 #endif
 
 static int airoha_alloc_gdm_port(struct airoha_eth *eth,
-				 struct device_node *np, int index)
+				 struct device_node *np)
 {
 	const __be32 *id_ptr = of_get_property(np, "reg", NULL);
 	struct airoha_gdm_port *port;
-	struct airoha_qdma *qdma;
 	struct net_device *dev;
 	int err, p;
 	u32 id;
@@ -3049,7 +3050,6 @@ static int airoha_alloc_gdm_port(struct
 		return -ENOMEM;
 	}
 
-	qdma = &eth->qdma[index % AIROHA_MAX_NUM_QDMA];
 	dev->netdev_ops = &airoha_netdev_ops;
 	dev->ethtool_ops = &airoha_ethtool_ops;
 	dev->max_mtu = AIROHA_MAX_MTU;
@@ -3061,7 +3061,6 @@ static int airoha_alloc_gdm_port(struct
 	dev->features |= dev->hw_features;
 	dev->vlan_features = dev->hw_features;
 	dev->dev.of_node = np;
-	dev->irq = qdma->irq_banks[0].irq;
 	SET_NETDEV_DEV(dev, eth->dev);
 
 	/* reserve hw queues for HTB offloading */
@@ -3082,7 +3081,7 @@ static int airoha_alloc_gdm_port(struct
 	port = netdev_priv(dev);
 	u64_stats_init(&port->stats.syncp);
 	spin_lock_init(&port->stats.lock);
-	port->qdma = qdma;
+	port->eth = eth;
 	port->dev = dev;
 	port->id = id;
 	eth->ports[p] = port;
@@ -3183,7 +3182,6 @@ static int airoha_probe(struct platform_
 	for (i = 0; i < ARRAY_SIZE(eth->qdma); i++)
 		airoha_qdma_start_napi(&eth->qdma[i]);
 
-	i = 0;
 	for_each_child_of_node(pdev->dev.of_node, np) {
 		if (!of_device_is_compatible(np, "airoha,eth-mac"))
 			continue;
@@ -3191,7 +3189,7 @@ static int airoha_probe(struct platform_
 		if (!of_device_is_available(np))
 			continue;
 
-		err = airoha_alloc_gdm_port(eth, np, i++);
+		err = airoha_alloc_gdm_port(eth, np);
 		if (err) {
 			of_node_put(np);
 			goto error_napi_stop;
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -533,6 +533,7 @@ struct airoha_qdma {
 
 struct airoha_gdm_port {
 	struct airoha_qdma *qdma;
+	struct airoha_eth *eth;
 	struct net_device *dev;
 	int id;
 
